src/Microsoft.AspNet.Server.Kestrel/Http/MessageBody.cs

```diff
@@ -111,29 +132,43 @@ public override void Consume()
         class ForContentLength : MessageBody
         {
             private readonly int _contentLength;
-            private int _neededLength;
+            private int _inputLength;
 
             public ForContentLength(bool keepAlive, int contentLength, FrameContext context)
                 : base(context)
             {
                 RequestKeepAlive = keepAlive;
                 _contentLength = contentLength;
-                _neededLength = _contentLength;
+                _inputLength = _contentLength;
             }
 
-            public override void Consume()
+            public override async Task<int> ReadAsyncImplementation(ArraySegment<byte> buffer, CancellationToken cancellationToken)
             {
                 var input = _context.SocketInput;
-                var consumeLength = Math.Min(_neededLength, input.Buffer.Count);
-                _neededLength -= consumeLength;
 
-                if (_neededLength != 0)
+                while (true)
                 {
-                    Intake(consumeLength);
-                }
-                else
-                {
-                    IntakeFin(consumeLength);
+                    var limit = Math.Min(buffer.Count, _inputLength);
+                    if (limit == 0)
+                    {
+                        return 0;
+                    }
+
+                    await input;
+
+                    var begin = input.ConsumingStart();
+                    int actual;
+                    var end = begin.CopyTo(buffer.Array, buffer.Offset, limit, out actual);
+                    _inputLength -= actual;
+                    input.ConsumingComplete(end, end);
+                    if (actual != 0)
+                    {
+                        return actual;
+                    }
+                    if (input.RemoteIntakeFin)
+                    {
+                        throw new Exception("Unexpected end of request content");
```

Comment:
I changed this to InvalidDataException. This one will throw back from Request.Body.Read/ReadAsync. Then it is caught where it's caught, worst-case disconnecting the connection when the frame's async processing function does it's finally {} clause